File: riccardo-negri/5/x.php
<?php

// Day 5 solution of Advent Of Code 2020 by Riccardo Negri
// First part answer: 991
// Second part answer: 534

const COL_BITS = 3; 
const ROW_BITS = 7; 

$index = 0;
if ($file = fopen("input.txt", "r")) {
    while(!feof($file)) {
        $line = fgets($file);
        $line = str_replace('F', '0', $line);
        $line = str_replace('B', '1', $line);
        $line = str_replace('L', '0', $line);
        $line = str_replace('R', '1', $line);
        $row = bindec(substr($line,0,7));
        $column = bindec(substr($line,7,10));
        $ID = $row * 8 + $column;
        $SeatsInfo[0][$index] = $row;
        $SeatsInfo[1][$index] = $column;
        $SeatsInfo[2][$index] = $ID;	
        $index++;
    }
    fclose($file);
}

echo "Max ID (First part answer): ".max($SeatsInfo[2])."\n";

foreach (range(0, 2**(COL_BITS)-1) as $column) {
    foreach (range(0, 2**(ROW_BITS)-1) as $row) {
        $ID = $row * 8 + $column;
        if (!in_array($ID, $SeatsInfo[2])) {
        	if (in_array(($ID-1), $SeatsInfo[2]) and in_array(($ID+1), $SeatsInfo[2])) {
        		$myID =  $ID;
        	}
        }
    }
}

echo "My ID (Second part answer): ".$myID."\n";

?>

File: riccardo-negri/10/x.py
#!/usr/bin/env python
# Day 10 solution of Advent Of Code 2020 by Riccardo Negri
# First part answer: 2046
# Second part answer: 1157018619904 (2*7*4*4*2*4*7*7*4*7*4*7*7*7*7*7*7)

numbers = []
with open('input.txt', 'r') as f:
		for line in f.readlines():
			numbers.append(int(line))

numbers.append(0)
numbers.append(max(numbers)+3)
numbers = sorted(numbers)

one_jolt = 0
three_jolt = 0
for index, num in enumerate(numbers):
	if index+1 < len(numbers):
		if numbers[index+1]-num == 1:
			one_jolt += 1
		elif numbers[index+1]-num == 3:
			three_jolt += 1

print("First part answer:  ", (one_jolt)*(three_jolt))

def number_of_possible_adapters_to_connect(index):
	count = 0
	for i in range(1,4):
		try:
			if numbers[index+i]-numbers[index] <= 3:
				count += 1
		except IndexError:
			return count
	return count

dictionary = {}
for index, num in enumerate(numbers):
	dictionary[num] = number_of_possible_adapters_to_connect(index)

# The input file has some fixed patterns.
# So I found out by hand the values for the three different patterns possible.
# The patterns are 1 2 1, 1 3 2 1, 1 3 3 2 1. To visualize them use print(dictionary).
# I don't care about the 1 because it doesn't generate any fork.
forks = 1
for index,num in enumerate(numbers):
	if dictionary[num] == 2:
		if dictionary[numbers[index-1]] == 3:
			if dictionary[numbers[index-2]] == 3:
				forks *= 7
			else:
				forks *= 4
		else:
			forks *= 2
print("Second part answer: ", forks)